### ISSUE TITLE
[clang] Misleading warning location for -Wimplicit-int-conversion

### ISSUE THREAD

**Issue by carlosgalvezp** [llvm/llvm-project#135136]
Consider this example:

```cpp
#include <cstdint>

std::uint16_t compute(std::uint16_t a, std::uint16_t b)
{
    return ((a & 0xFF) << 3) | (b & 0x0F);
}
```

Enabling `-Wimplicit-int-conversion`, Clang prints:

```
source>:5:30: warning: implicit conversion loses integer precision: 'int' to 'std::uint16_t' (aka 'unsigned short') [-Wimplicit-int-conversion]
    5 |     return ((a & 0xFF) << 3) | (b & 0x0F);
      |     ~~~~~~ ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
```

[Godbolt](https://godbolt.org/z/qf6nYnK9M)

This is very confusing, because the caret `^` is pointing at the `|` sign, and that's not where the problem lies. Therefore it's quite misleading and people may spend a lot of time figuring out what's going on.

The implicit conversion is happening because the result of these operations is `int`, and the return type of the function is `std::uint16_t`. What the user needs to do to fix the warning is to cast the result to `std::uint16_t`. But this is not at all obvious from the warning message, especially if beginner users aren't familiar with the rules of integer promotion/usual arithmetic conversions.

Could this warning be improved to more clearly highlight/pinpoint what the problem is? 

Thanks!

**Comment by firewave** [comment]
Looks like the `int` version of #93288.

**Comment by carlosgalvezp** [comment]
Pretty much, yes! Since it's two separate warnings I believe it make sense to keep them as separate tickets.

**Comment by zygoloid** [comment]
Two things are being underlined here: the `return` is the place where the conversion is being performed, and the expression to its right is the thing being converted. This is presumably using Clang's `getExprLoc` mechanism for the "thing being converted" part to point at the root of the AST for that expression, which in this case is the `|` operator. That seems like a mistake to me, since it suggests a problem with that operator in particular. It'd be better to use `getBeginLoc` rather than `getExprLoc` when producing diagnostics [here](https://github.com/llvm/llvm-project/blob/c893f1d18da6547027648f44592f61c08570d80f/clang/lib/Sema/SemaChecking.cpp#L11358) -- but someone will need to check some cases to make sure this is an improvement for the various ways this function can be called. That'd give:

```
<source>:5:30: warning: implicit conversion loses integer precision: 'int' to 'std::uint16_t' (aka 'unsigned short') [-Wimplicit-int-conversion]
    5 |     return ((a & 0xFF) << 3) | (b & 0x0F);
      |     ~~~~~~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

which seems much less confusing to me.

Separately, isn't the warning simply wrong here? `a & 0xFF` is 8 bits wide and non-negative, `(a & 0xFF) << 3` is 11 bits wide and non-negative, `b & 0x0F` is 4 bits wide and non-negative, so `((a & 0xFF) << 3) | (b & 0x0F)` is 11 bits wide and non-negative, so we should not be warning about a loss of precision when converting that expression to `uint16_t`. It seems like a bug in `TryGetExprRange` if we can't work that out and avoid diagnosing this!
